File: ios/DesMoinesInsider/Services/AnalyticsService.swift
import Foundation
import os
import CryptoKit

/// EMITS NOTHING. This is a stub, and every track* call below ends at a debug
/// log line that release builds do not keep.
///
/// The previous docstring described it as "a lightweight analytics facade that
/// wraps Firebase Analytics", which reads as working code with Firebase behind
/// it. There is no Firebase dependency, no GoogleService-Info.plist, and every
/// Analytics.logEvent call is commented out. Anyone reading a funnel report and
/// seeing no iOS data would reasonably conclude iOS users are not converting,
/// rather than that iOS never reported (IOS-AUDIT-FEAT-030 AC5, XPLAT-004 AC5).
///
/// ANDROID DOES EMIT, through a real Firebase sink, and its own docstring says
/// "Event names mirror iOS verbatim". The names do match; the data does not.
/// Any cross-platform comparison built on that is measuring Android against
/// zero.
///
/// The consent gate below is real and is kept: when this is activated, the gate
/// must already be in the path rather than be added afterwards.
///
/// To activate (XPLAT-004 AC2 decides whether Firebase is the right target at
/// all - the web surface uses Sentry, and routing both mobile apps through the
/// existing log-error function would unify them instead):
///   1. Add Firebase SPM package to project.yml
///   2. Add GoogleService-Info.plist to Resources
///   3. Uncomment Firebase imports and calls below
///   4. Update PrivacyInfo.xcprivacy with analytics data collection
@MainActor
final class AnalyticsService {
    static let shared = AnalyticsService()

    private init() {}

    // MARK: - Screen Views

    func trackScreenView(_ screenName: String) {
        guard ConsentService.shared.analyticsConsent else { return }
        // Analytics.logEvent(AnalyticsEventScreenView, parameters: [AnalyticsParameterScreenName: screenName])
        AppLogger.general.debug("Analytics: screen_view — \(screenName)")
    }

    // MARK: - Events

    func trackEvent(_ name: String, parameters: [String: Any]? = nil) {
        guard ConsentService.shared.analyticsConsent else { return }
        // Analytics.logEvent(name, parameters: parameters)
        AppLogger.general.debug("Analytics: \(name)")
    }

    // MARK: - Pre-defined Events

    func trackAppOpen() {
        trackEvent("app_open")
    }

    func trackViewEvent(eventId: String, category: String?) {
        trackEvent("view_event", parameters: [
            "event_id": eventId,
            "category": category ?? "unknown",
        ])
    }

    func trackViewRestaurant(restaurantId: String, cuisine: String?) {
        trackEvent("view_restaurant", parameters: [
            "restaurant_id": restaurantId,
            "cuisine": cuisine ?? "unknown",
        ])
    }

    func trackSearch(query: String, resultCount: Int) {
        trackEvent("search", parameters: [
            "query_length": query.count,
            "result_count": resultCount,
        ])
    }

    func trackSaveFavorite(itemType: String, itemId: String) {
        trackEvent("save_favorite", parameters: [
            "item_type": itemType,
            "item_id": itemId,
        ])
    }

    func trackSubscriptionView() {
        trackEvent("subscription_view")
    }

    func trackPurchaseStart(productId: String) {
        trackEvent("purchase_start", parameters: ["product_id": productId])
    }

    func trackPurchaseComplete(productId: String) {
        trackEvent("purchase_complete", parameters: ["product_id": productId])
    }

    // MARK: - Contextual Paywall (IOS-SUB-010)
    //
    // Every paywall lifecycle event carries the `context` id (e.g.
    // "unlimited_favorites", "trip_planner") so per-surface conversion can be
    // measured — which upsell moments convert and which don't.

    func trackPaywallPresented(context: String, tier: String) {
        trackEvent("paywall_present", parameters: ["context": context, "tier": tier])
    }

    func trackPaywallDismissed(context: String) {
        trackEvent("paywall_dismiss", parameters: ["context": context])
    }

    func trackPaywallPurchaseStart(context: String, productId: String) {
        trackEvent("paywall_purchase_start", parameters: ["context": context, "product_id": productId])
    }

    func trackPaywallPurchaseComplete(context: String, productId: String) {
        trackEvent("paywall_purchase_complete", parameters: ["context": context, "product_id": productId])
    }

    func trackPaywallRestore(context: String) {
        trackEvent("paywall_restore", parameters: ["context": context])
    }

    // MARK: - Onboarding & soft paywall (IOS-SUB-013)
    //
    // `action`: shown / start_tapped / skipped — for first-session trial funnel.
    // `source`: which engagement threshold fired the post-onboarding soft paywall.

    func trackOnboardingTrial(action: String) {
        trackEvent("onboarding_trial", parameters: ["action": action, "cohort": "onboarding"])
    }

    func trackSoftPaywall(source: String) {
        trackEvent("soft_paywall_present", parameters: ["source": source])
    }

    // MARK: - Win-back / renewal (IOS-SUB-014)

    /// `state`: expiring_soon / billing_retry / grace / expired.
    func trackRenewalBanner(action: String, state: String) {
        trackEvent("renewal_banner", parameters: ["action": action, "state": state])
    }

    func trackOfferCodeRedeem(action: String) {
        trackEvent("offer_code_redeem", parameters: ["action": action])
    }

    // MARK: - User Properties

    func setUserId(_ userId: String?) {
        guard ConsentService.shared.analyticsConsent else { return }
        if let userId {
            // SHA-256 the UUID, then truncate. The prior code hex-encoded the raw
            // UTF-8 bytes — a reversible encoding of the literal UUID, not a hash
            // (IOS-AUDIT-SEC-004). A truncated digest is not derivable back to the
            // source id.
            let digest = SHA256.hash(data: Data(userId.utf8))
            let hex = digest.map { String(format: "%02x", $0) }.joined()
            let pseudonymousId = String(hex.prefix(16))
            // Analytics.setUserID(pseudonymousId)
            AppLogger.general.debug("Analytics: user set \(pseudonymousId)")
        } else {
            // Analytics.setUserID(nil)
        }
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/util/AnalyticsService.kt
package com.desmoines.aipulse.util

import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Sink that actually records an analytics event. Abstracted so AnalyticsService's
 * consent-gating logic is unit-testable without the Firebase SDK / android.os.Bundle.
 */
interface AnalyticsSink {
    fun send(name: String, params: Map<String, String>)
    fun setCollectionEnabled(enabled: Boolean)
}

/** Firebase-backed [AnalyticsSink]. Not unit-tested (thin Bundle adapter). */
class FirebaseAnalyticsSink(
    private val firebaseAnalytics: FirebaseAnalytics,
) : AnalyticsSink {
    override fun send(name: String, params: Map<String, String>) {
        val bundle = Bundle().apply { params.forEach { (k, v) -> putString(k, v) } }
        firebaseAnalytics.logEvent(name, bundle)
    }

    override fun setCollectionEnabled(enabled: Boolean) {
        firebaseAnalytics.setAnalyticsCollectionEnabled(enabled)
    }
}

/**
 * Consent-gated analytics facade. Mirrors iOS AnalyticsService.swift.
 *
 * Non-essential events are dropped unless ConsentService.analyticsConsent is true.
 * Essential telemetry (crash/auth-error breadcrumbs) is never gated. No PII —
 * event names + structural parameters only.
 *
 * Event NAMES mirror iOS verbatim. The DATA does not: iOS AnalyticsService is
 * a stub whose every Firebase call is commented out, so it emits nothing at
 * all. This class is the only mobile surface reporting. Do not read a
 * cross-platform funnel from these events - it compares Android against zero,
 * and an absent iOS line means iOS never reported, not that iOS users did not
 * convert (XPLAT-004 AC5).
 */
@Singleton
class AnalyticsService @Inject constructor(
    private val sink: AnalyticsSink,
    private val consentService: ConsentService,
) {
    init {
        // Reflect the user's current choice in Firebase collection at startup.
        sink.setCollectionEnabled(consentService.analyticsConsent.value)
    }

    private val hasConsent: Boolean get() = consentService.analyticsConsent.value

    /** Re-apply collection state after a consent change. */
    fun applyConsent() {
        sink.setCollectionEnabled(hasConsent)
    }

    /**
     * Log [name] with [params]. Non-essential events are dropped without consent;
     * [essential] events (breadcrumbs) always send.
     */
    fun logEvent(name: String, params: Map<String, String> = emptyMap(), essential: Boolean = false) {
        if (!essential && !hasConsent) return
        sink.send(name, params)
    }

    fun logAppOpen() = logEvent(EVENT_APP_OPEN)

    fun logScreen(screen: String) = logEvent(EVENT_SCREEN_VIEW, mapOf(PARAM_SCREEN to screen))

    fun logSearch(queryLength: Int, resultCount: Int) = logEvent(
        EVENT_SEARCH,
        mapOf(PARAM_QUERY_LENGTH to queryLength.toString(), PARAM_RESULT_COUNT to resultCount.toString()),
    )

    fun logSaveFavorite(itemType: String) =
        logEvent(EVENT_SAVE_FAVORITE, mapOf(PARAM_ITEM_TYPE to itemType))

    fun logPaywallPresent(context: String) =
        logEvent(EVENT_PAYWALL_PRESENT, mapOf(PARAM_CONTEXT to context))

    fun logPaywallDismiss(context: String) =
        logEvent(EVENT_PAYWALL_DISMISS, mapOf(PARAM_CONTEXT to context))

    fun logPaywallPurchaseStart(context: String) =
        logEvent(EVENT_PAYWALL_PURCHASE_START, mapOf(PARAM_CONTEXT to context))

    companion object {
        const val EVENT_APP_OPEN = "app_open"
        const val EVENT_SCREEN_VIEW = "screen_view"
        const val EVENT_SEARCH = "search"
        const val EVENT_SAVE_FAVORITE = "save_favorite"
        const val EVENT_PAYWALL_PRESENT = "paywall_present"
        const val EVENT_PAYWALL_DISMISS = "paywall_dismiss"
        const val EVENT_PAYWALL_PURCHASE_START = "paywall_purchase_start"

        const val PARAM_SCREEN = "screen"
        const val PARAM_QUERY_LENGTH = "query_length"
        const val PARAM_RESULT_COUNT = "result_count"
        const val PARAM_ITEM_TYPE = "item_type"
        const val PARAM_CONTEXT = "context"
    }
}
